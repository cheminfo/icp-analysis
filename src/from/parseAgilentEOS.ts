import XLSX from 'xlsx';

export function parseAgilentEOS(binary: ArrayBuffer | Uint8Array) {
  const workbook = XLSX.read(binary, { type: 'array' });

  // summary is on Sheet2
  const matrix: (string | number)[][] = XLSX.utils.sheet_to_json(
    workbook.Sheets[workbook.SheetNames[1]],
    {
      header: 1,
    },
  );
  const header = parseHeader(matrix[0]);
  const samples = [];
  for (let line of matrix.slice(1).filter((line) => line[2])) {
    const sample = { elements: [] as any, reference: line[2] };
    for (let i = 3; i < header.length; i++) {
      const result = JSON.parse(JSON.stringify(header[i]));
      if (result.quantity.value === 'undefined') {
        continue;
      }
      result.quantity.value = parseFloat(`${line[i]}`);
      sample.elements.push(result);
    }
    samples.push(sample);
  }
  return samples;
}

function parseHeader(header: any) {
  return header.map((item: string) => {
    if (item.includes('.')) {
      const parts = item.split(' ');
      return {
        element: parts[0],
        wavelength: {
          value: Number(parts[1]),
          units: parts[2],
        },
        quantity: {
          value: undefined,
          units: parts[3] === 'ppm' ? 'mg/l' : parts[3],
        },
      };
    }
    return { label: item };
  });
}
