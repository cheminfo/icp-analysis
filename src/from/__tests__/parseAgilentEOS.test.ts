import { readFileSync } from 'fs';
import { join } from 'path';

import { parseAgilentEOS } from '../parseAgilentEOS';

describe('parseAgilentEOS', () => {
  it('no dilution', () => {
    const binary = readFileSync(join(__dirname, 'data/agilent.xlsx'));
    const results = parseAgilentEOS(binary);
    expect(results).toHaveLength(19);
    expect(results[17]).toStrictEqual({
      elements: [
        {
          element: 'Al',
          wavelength: { value: 308.215, units: 'nm' },
          experimentalConcentration: { units: 'mg/l', value: 0.12 },
        },
        {
          element: 'Al',
          wavelength: { value: 309.271, units: 'nm' },
          experimentalConcentration: { units: 'mg/l', value: 0.1 },
        },
        {
          element: 'Al',
          wavelength: { value: 396.152, units: 'nm' },
          experimentalConcentration: { units: 'mg/l', value: 0.12 },
        },
        {
          element: 'Cu',
          wavelength: { value: 213.598, units: 'nm' },
          experimentalConcentration: { units: 'mg/l', value: 2.67 },
        },
        {
          element: 'Cu',
          wavelength: { value: 324.754, units: 'nm' },
          experimentalConcentration: { units: 'mg/l', value: 2.66 },
        },
        {
          element: 'Cu',
          wavelength: { value: 327.395, units: 'nm' },
          experimentalConcentration: { units: 'mg/l', value: 2.65 },
        },
      ],
      reference: 'patiny_TEST_F3',
    });
    expect(results).toMatchSnapshot();
  });
  it('dilution', () => {
    const binary = readFileSync(join(__dirname, 'data/agilent.xlsx'));
    const results = parseAgilentEOS(binary, {
      dilution: { factor: 10, solvent: 'Nitric acid 2%' },
    });
    expect(results).toHaveLength(19);
    expect(results[17].elements[0]).toStrictEqual({
      element: 'Al',
      wavelength: { value: 308.215, units: 'nm' },
      experimentalConcentration: { units: 'mg/l', value: 0.12 },
      sampleConcentration: { value: 1.2, units: 'mg/l' },
      dilution: { factor: 10, solvent: 'Nitric acid 2%' },
    });
  });
});
