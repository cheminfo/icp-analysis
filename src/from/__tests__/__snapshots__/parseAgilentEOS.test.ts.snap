// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseAgilentEOS 1`] = `
Array [
  Object {
    "elements": Array [
      Object {
        "element": "Al",
        "quantity": Object {
          "units": "mg/l",
          "value": 0,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 308.215,
        },
      },
      Object {
        "element": "Al",
        "quantity": Object {
          "units": "mg/l",
          "value": 0,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 309.271,
        },
      },
      Object {
        "element": "Al",
        "quantity": Object {
          "units": "mg/l",
          "value": 0,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 396.152,
        },
      },
      Object {
        "element": "Cu",
        "quantity": Object {
          "units": "mg/l",
          "value": 0,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 213.598,
        },
      },
      Object {
        "element": "Cu",
        "quantity": Object {
          "units": "mg/l",
          "value": 0,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 324.754,
        },
      },
      Object {
        "element": "Cu",
        "quantity": Object {
          "units": "mg/l",
          "value": 0,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 327.395,
        },
      },
    ],
    "reference": "Blank",
  },
  Object {
    "elements": Array [
      Object {
        "element": "Al",
        "quantity": Object {
          "units": "mg/l",
          "value": 0.01,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 308.215,
        },
      },
      Object {
        "element": "Al",
        "quantity": Object {
          "units": "mg/l",
          "value": 0.01,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 309.271,
        },
      },
      Object {
        "element": "Al",
        "quantity": Object {
          "units": "mg/l",
          "value": 0.01,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 396.152,
        },
      },
      Object {
        "element": "Cu",
        "quantity": Object {
          "units": "mg/l",
          "value": NaN,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 213.598,
        },
      },
      Object {
        "element": "Cu",
        "quantity": Object {
          "units": "mg/l",
          "value": NaN,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 324.754,
        },
      },
      Object {
        "element": "Cu",
        "quantity": Object {
          "units": "mg/l",
          "value": NaN,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 327.395,
        },
      },
    ],
    "reference": "Standard 1",
  },
  Object {
    "elements": Array [
      Object {
        "element": "Al",
        "quantity": Object {
          "units": "mg/l",
          "value": 0.05,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 308.215,
        },
      },
      Object {
        "element": "Al",
        "quantity": Object {
          "units": "mg/l",
          "value": 0.05,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 309.271,
        },
      },
      Object {
        "element": "Al",
        "quantity": Object {
          "units": "mg/l",
          "value": 0.05,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 396.152,
        },
      },
      Object {
        "element": "Cu",
        "quantity": Object {
          "units": "mg/l",
          "value": NaN,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 213.598,
        },
      },
      Object {
        "element": "Cu",
        "quantity": Object {
          "units": "mg/l",
          "value": NaN,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 324.754,
        },
      },
      Object {
        "element": "Cu",
        "quantity": Object {
          "units": "mg/l",
          "value": NaN,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 327.395,
        },
      },
    ],
    "reference": "Standard 2",
  },
  Object {
    "elements": Array [
      Object {
        "element": "Al",
        "quantity": Object {
          "units": "mg/l",
          "value": 0.1,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 308.215,
        },
      },
      Object {
        "element": "Al",
        "quantity": Object {
          "units": "mg/l",
          "value": 0.1,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 309.271,
        },
      },
      Object {
        "element": "Al",
        "quantity": Object {
          "units": "mg/l",
          "value": 0.1,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 396.152,
        },
      },
      Object {
        "element": "Cu",
        "quantity": Object {
          "units": "mg/l",
          "value": NaN,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 213.598,
        },
      },
      Object {
        "element": "Cu",
        "quantity": Object {
          "units": "mg/l",
          "value": NaN,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 324.754,
        },
      },
      Object {
        "element": "Cu",
        "quantity": Object {
          "units": "mg/l",
          "value": NaN,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 327.395,
        },
      },
    ],
    "reference": "Standard 3",
  },
  Object {
    "elements": Array [
      Object {
        "element": "Al",
        "quantity": Object {
          "units": "mg/l",
          "value": 0.5,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 308.215,
        },
      },
      Object {
        "element": "Al",
        "quantity": Object {
          "units": "mg/l",
          "value": 0.5,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 309.271,
        },
      },
      Object {
        "element": "Al",
        "quantity": Object {
          "units": "mg/l",
          "value": 0.5,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 396.152,
        },
      },
      Object {
        "element": "Cu",
        "quantity": Object {
          "units": "mg/l",
          "value": NaN,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 213.598,
        },
      },
      Object {
        "element": "Cu",
        "quantity": Object {
          "units": "mg/l",
          "value": NaN,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 324.754,
        },
      },
      Object {
        "element": "Cu",
        "quantity": Object {
          "units": "mg/l",
          "value": NaN,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 327.395,
        },
      },
    ],
    "reference": "Standard 4",
  },
  Object {
    "elements": Array [
      Object {
        "element": "Al",
        "quantity": Object {
          "units": "mg/l",
          "value": 1,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 308.215,
        },
      },
      Object {
        "element": "Al",
        "quantity": Object {
          "units": "mg/l",
          "value": 1,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 309.271,
        },
      },
      Object {
        "element": "Al",
        "quantity": Object {
          "units": "mg/l",
          "value": 1,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 396.152,
        },
      },
      Object {
        "element": "Cu",
        "quantity": Object {
          "units": "mg/l",
          "value": NaN,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 213.598,
        },
      },
      Object {
        "element": "Cu",
        "quantity": Object {
          "units": "mg/l",
          "value": NaN,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 324.754,
        },
      },
      Object {
        "element": "Cu",
        "quantity": Object {
          "units": "mg/l",
          "value": NaN,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 327.395,
        },
      },
    ],
    "reference": "Standard 5",
  },
  Object {
    "elements": Array [
      Object {
        "element": "Al",
        "quantity": Object {
          "units": "mg/l",
          "value": 5,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 308.215,
        },
      },
      Object {
        "element": "Al",
        "quantity": Object {
          "units": "mg/l",
          "value": 5,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 309.271,
        },
      },
      Object {
        "element": "Al",
        "quantity": Object {
          "units": "mg/l",
          "value": 5,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 396.152,
        },
      },
      Object {
        "element": "Cu",
        "quantity": Object {
          "units": "mg/l",
          "value": NaN,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 213.598,
        },
      },
      Object {
        "element": "Cu",
        "quantity": Object {
          "units": "mg/l",
          "value": NaN,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 324.754,
        },
      },
      Object {
        "element": "Cu",
        "quantity": Object {
          "units": "mg/l",
          "value": NaN,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 327.395,
        },
      },
    ],
    "reference": "Standard 6",
  },
  Object {
    "elements": Array [
      Object {
        "element": "Al",
        "quantity": Object {
          "units": "mg/l",
          "value": 10,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 308.215,
        },
      },
      Object {
        "element": "Al",
        "quantity": Object {
          "units": "mg/l",
          "value": 10,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 309.271,
        },
      },
      Object {
        "element": "Al",
        "quantity": Object {
          "units": "mg/l",
          "value": 10,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 396.152,
        },
      },
      Object {
        "element": "Cu",
        "quantity": Object {
          "units": "mg/l",
          "value": NaN,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 213.598,
        },
      },
      Object {
        "element": "Cu",
        "quantity": Object {
          "units": "mg/l",
          "value": NaN,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 324.754,
        },
      },
      Object {
        "element": "Cu",
        "quantity": Object {
          "units": "mg/l",
          "value": NaN,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 327.395,
        },
      },
    ],
    "reference": "Standard 7",
  },
  Object {
    "elements": Array [
      Object {
        "element": "Al",
        "quantity": Object {
          "units": "mg/l",
          "value": NaN,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 308.215,
        },
      },
      Object {
        "element": "Al",
        "quantity": Object {
          "units": "mg/l",
          "value": NaN,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 309.271,
        },
      },
      Object {
        "element": "Al",
        "quantity": Object {
          "units": "mg/l",
          "value": NaN,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 396.152,
        },
      },
      Object {
        "element": "Cu",
        "quantity": Object {
          "units": "mg/l",
          "value": 0.01,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 213.598,
        },
      },
      Object {
        "element": "Cu",
        "quantity": Object {
          "units": "mg/l",
          "value": 0.01,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 324.754,
        },
      },
      Object {
        "element": "Cu",
        "quantity": Object {
          "units": "mg/l",
          "value": 0.01,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 327.395,
        },
      },
    ],
    "reference": "Standard 8",
  },
  Object {
    "elements": Array [
      Object {
        "element": "Al",
        "quantity": Object {
          "units": "mg/l",
          "value": NaN,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 308.215,
        },
      },
      Object {
        "element": "Al",
        "quantity": Object {
          "units": "mg/l",
          "value": NaN,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 309.271,
        },
      },
      Object {
        "element": "Al",
        "quantity": Object {
          "units": "mg/l",
          "value": NaN,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 396.152,
        },
      },
      Object {
        "element": "Cu",
        "quantity": Object {
          "units": "mg/l",
          "value": 0.05,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 213.598,
        },
      },
      Object {
        "element": "Cu",
        "quantity": Object {
          "units": "mg/l",
          "value": 0.05,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 324.754,
        },
      },
      Object {
        "element": "Cu",
        "quantity": Object {
          "units": "mg/l",
          "value": 0.05,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 327.395,
        },
      },
    ],
    "reference": "Standard 9",
  },
  Object {
    "elements": Array [
      Object {
        "element": "Al",
        "quantity": Object {
          "units": "mg/l",
          "value": NaN,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 308.215,
        },
      },
      Object {
        "element": "Al",
        "quantity": Object {
          "units": "mg/l",
          "value": NaN,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 309.271,
        },
      },
      Object {
        "element": "Al",
        "quantity": Object {
          "units": "mg/l",
          "value": NaN,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 396.152,
        },
      },
      Object {
        "element": "Cu",
        "quantity": Object {
          "units": "mg/l",
          "value": 0.1,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 213.598,
        },
      },
      Object {
        "element": "Cu",
        "quantity": Object {
          "units": "mg/l",
          "value": 0.1,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 324.754,
        },
      },
      Object {
        "element": "Cu",
        "quantity": Object {
          "units": "mg/l",
          "value": 0.1,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 327.395,
        },
      },
    ],
    "reference": "Standard 10",
  },
  Object {
    "elements": Array [
      Object {
        "element": "Al",
        "quantity": Object {
          "units": "mg/l",
          "value": NaN,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 308.215,
        },
      },
      Object {
        "element": "Al",
        "quantity": Object {
          "units": "mg/l",
          "value": NaN,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 309.271,
        },
      },
      Object {
        "element": "Al",
        "quantity": Object {
          "units": "mg/l",
          "value": NaN,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 396.152,
        },
      },
      Object {
        "element": "Cu",
        "quantity": Object {
          "units": "mg/l",
          "value": 0.5,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 213.598,
        },
      },
      Object {
        "element": "Cu",
        "quantity": Object {
          "units": "mg/l",
          "value": 0.5,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 324.754,
        },
      },
      Object {
        "element": "Cu",
        "quantity": Object {
          "units": "mg/l",
          "value": 0.5,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 327.395,
        },
      },
    ],
    "reference": "Standard 11",
  },
  Object {
    "elements": Array [
      Object {
        "element": "Al",
        "quantity": Object {
          "units": "mg/l",
          "value": NaN,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 308.215,
        },
      },
      Object {
        "element": "Al",
        "quantity": Object {
          "units": "mg/l",
          "value": NaN,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 309.271,
        },
      },
      Object {
        "element": "Al",
        "quantity": Object {
          "units": "mg/l",
          "value": NaN,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 396.152,
        },
      },
      Object {
        "element": "Cu",
        "quantity": Object {
          "units": "mg/l",
          "value": 1,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 213.598,
        },
      },
      Object {
        "element": "Cu",
        "quantity": Object {
          "units": "mg/l",
          "value": 1,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 324.754,
        },
      },
      Object {
        "element": "Cu",
        "quantity": Object {
          "units": "mg/l",
          "value": 1,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 327.395,
        },
      },
    ],
    "reference": "Standard 12",
  },
  Object {
    "elements": Array [
      Object {
        "element": "Al",
        "quantity": Object {
          "units": "mg/l",
          "value": NaN,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 308.215,
        },
      },
      Object {
        "element": "Al",
        "quantity": Object {
          "units": "mg/l",
          "value": NaN,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 309.271,
        },
      },
      Object {
        "element": "Al",
        "quantity": Object {
          "units": "mg/l",
          "value": NaN,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 396.152,
        },
      },
      Object {
        "element": "Cu",
        "quantity": Object {
          "units": "mg/l",
          "value": 5,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 213.598,
        },
      },
      Object {
        "element": "Cu",
        "quantity": Object {
          "units": "mg/l",
          "value": 5,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 324.754,
        },
      },
      Object {
        "element": "Cu",
        "quantity": Object {
          "units": "mg/l",
          "value": 5,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 327.395,
        },
      },
    ],
    "reference": "Standard 13",
  },
  Object {
    "elements": Array [
      Object {
        "element": "Al",
        "quantity": Object {
          "units": "mg/l",
          "value": NaN,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 308.215,
        },
      },
      Object {
        "element": "Al",
        "quantity": Object {
          "units": "mg/l",
          "value": NaN,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 309.271,
        },
      },
      Object {
        "element": "Al",
        "quantity": Object {
          "units": "mg/l",
          "value": NaN,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 396.152,
        },
      },
      Object {
        "element": "Cu",
        "quantity": Object {
          "units": "mg/l",
          "value": 10,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 213.598,
        },
      },
      Object {
        "element": "Cu",
        "quantity": Object {
          "units": "mg/l",
          "value": 10,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 324.754,
        },
      },
      Object {
        "element": "Cu",
        "quantity": Object {
          "units": "mg/l",
          "value": 10,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 327.395,
        },
      },
    ],
    "reference": "Standard 14",
  },
  Object {
    "elements": Array [
      Object {
        "element": "Al",
        "quantity": Object {
          "units": "mg/l",
          "value": 0.01,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 308.215,
        },
      },
      Object {
        "element": "Al",
        "quantity": Object {
          "units": "mg/l",
          "value": -0.01,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 309.271,
        },
      },
      Object {
        "element": "Al",
        "quantity": Object {
          "units": "mg/l",
          "value": 0.01,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 396.152,
        },
      },
      Object {
        "element": "Cu",
        "quantity": Object {
          "units": "mg/l",
          "value": 0.6,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 213.598,
        },
      },
      Object {
        "element": "Cu",
        "quantity": Object {
          "units": "mg/l",
          "value": 0.59,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 324.754,
        },
      },
      Object {
        "element": "Cu",
        "quantity": Object {
          "units": "mg/l",
          "value": 0.59,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 327.395,
        },
      },
    ],
    "reference": "patiny_TEST_F1",
  },
  Object {
    "elements": Array [
      Object {
        "element": "Al",
        "quantity": Object {
          "units": "mg/l",
          "value": 0.17,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 308.215,
        },
      },
      Object {
        "element": "Al",
        "quantity": Object {
          "units": "mg/l",
          "value": 0.15,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 309.271,
        },
      },
      Object {
        "element": "Al",
        "quantity": Object {
          "units": "mg/l",
          "value": 0.17,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 396.152,
        },
      },
      Object {
        "element": "Cu",
        "quantity": Object {
          "units": "mg/l",
          "value": 2.6,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 213.598,
        },
      },
      Object {
        "element": "Cu",
        "quantity": Object {
          "units": "mg/l",
          "value": 2.58,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 324.754,
        },
      },
      Object {
        "element": "Cu",
        "quantity": Object {
          "units": "mg/l",
          "value": 2.57,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 327.395,
        },
      },
    ],
    "reference": "patiny_TEST_F2",
  },
  Object {
    "elements": Array [
      Object {
        "element": "Al",
        "quantity": Object {
          "units": "mg/l",
          "value": 0.12,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 308.215,
        },
      },
      Object {
        "element": "Al",
        "quantity": Object {
          "units": "mg/l",
          "value": 0.1,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 309.271,
        },
      },
      Object {
        "element": "Al",
        "quantity": Object {
          "units": "mg/l",
          "value": 0.12,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 396.152,
        },
      },
      Object {
        "element": "Cu",
        "quantity": Object {
          "units": "mg/l",
          "value": 2.67,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 213.598,
        },
      },
      Object {
        "element": "Cu",
        "quantity": Object {
          "units": "mg/l",
          "value": 2.66,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 324.754,
        },
      },
      Object {
        "element": "Cu",
        "quantity": Object {
          "units": "mg/l",
          "value": 2.65,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 327.395,
        },
      },
    ],
    "reference": "patiny_TEST_F3",
  },
  Object {
    "elements": Array [
      Object {
        "element": "Al",
        "quantity": Object {
          "units": "mg/l",
          "value": -0.02,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 308.215,
        },
      },
      Object {
        "element": "Al",
        "quantity": Object {
          "units": "mg/l",
          "value": -0.04,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 309.271,
        },
      },
      Object {
        "element": "Al",
        "quantity": Object {
          "units": "mg/l",
          "value": -0.02,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 396.152,
        },
      },
      Object {
        "element": "Cu",
        "quantity": Object {
          "units": "mg/l",
          "value": 17.15,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 213.598,
        },
      },
      Object {
        "element": "Cu",
        "quantity": Object {
          "units": "mg/l",
          "value": 16.91,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 324.754,
        },
      },
      Object {
        "element": "Cu",
        "quantity": Object {
          "units": "mg/l",
          "value": 16.94,
        },
        "wavelength": Object {
          "units": "nm",
          "value": 327.395,
        },
      },
    ],
    "reference": "patiny_TEST_F4",
  },
]
`;
